### PR TITLE
Improve telnet-mode handling for softmodems

### LINKED ISSUE #581: Telnet support enablement for "softmodem" serial ports
I was often asked why DOS terminal programs would fail to download files from BBSes when using DOSBox (e.g. using traditional serial/modem transfer protocols like XMODEM, YMODEM, ZMODEM). I'd seen it reported and confirmed that when a "serialX" port is configured in a DOSBox .conf as device of type "modem", that it does indeed, by default, not support Telnet. It may appear to support Telnet as it'll connect to Telnet servers, but it won't actually escape sent-IAC characters, unescape received-IAC characters or recognize or respond to Telnet commands. The "nullmodem" devicetype does support a "telnet:1" parameter to enable Telnet support, but that's not really helpful to those attempting to use DOS terminals to connect to Telnet BBSes on the Internet.

I looked through the DOSBox Wiki and Manual and really could find nothing about enabling Telnet support for "modem" (softmodem) serial devices. So I looked through the DOSBox source code and foudn this little bit of code in softmodem.cpp:

```
else if (strstr(cmdbuf,"NET1")) {
                telnetmode = true;
                SendRes(ResOK);
                return;
        }
```
That's some interesting code there. So if the 4 characters "NET1" appear **_anywhere_** in _any_ AT command received by this softmodem, it'll enable Telnet support and then immediately stop parsing the command and return "OK". That's pretty terrible AT command parsing there (what if "NET1" happens to be in the hostname you're trying to connect to?), but whatever. I tried typing "ATNET1", and indeed it works: the next connection using "ATD..." did indeed support Telnet.

So...
1. Please document this ATNET0/1 command (1 to turn Telnet support on 0, to turn if off).
2. Please fix the AT command parser: `strstr()` is almost never what you actually want to use.
3. Please support the "telnet:1" option in the .conf file to enable telnet suport for "modem" devices by default.

Happy to discuss further,

Thanks, BBS users and sysops thank you.

### PR DESCRIPTION
**Pull request by kcgen**
This PR:
- Prefixes the telnet-mode command with **+**: `+NET1` and `+NET0`, per the Hayes multi-character command specification.
- Moves parsing of these telnet-mode commands into the AT-command processing loop, so they can be processed like any other sequential command.
- Retains telnet-mode as set by the user, and does not (constantly) disable it when resetting the modem (ie: ATZ)
- Allows telnet-mode to be enabled from the get-go by adding `telnet:1` to the serial port's conf-file line.
- README is updated with the above changes.

Fixes #581 
To-be-confirmed by @rswindell (thank you for reporting these short-comings!)

### CODE REVIEW

**Comment by kcgen** [comment]
Tested extensively; here's the conf-file option in action, followed by a handful of AT-command testing showing mixed-parsing (and a deliberate failure in red):

## Configuation telnet:1 control



## AT-command parsing control



BBS connections and transfers continue to work as-expected with tenet-mode enabled.

@dreamer , given it's passing all of my testing, I propose we move ahead w/ review and merging.

**Comment by rswindell** [comment]
I haven't tested yet, but plan to.

Out of curiosity, does the DOSbox softmodem AT command parser support the "echo" option (e.g. ATE1 to enable, ATE0 to disable echo)? This would be the one reason I could think that the previous "NET0/1" option would collide with standard modem commands: The "E" command is the equivalent of "E0" (echo off), so "NET0" or "NET1" should have turned echo off (the N and T commands, along with the '0' or '1' value would be ignored if unsupported). It's an unfortunate backward-incompatible change, I but think the move to "!NET0/1" was necessary for true modem-like behavior.

Traditionally, multiple character modem commands were prefixed with '+' (e.g. "AT+FCLASS=2.0"), but I can't think of a reason why the '!' prefix should not work just as well.

**Comment by kcgen** [comment]
Thanks for extra insight @rswindell and lending your hand at testing it (you being the author of SynchroNet and prior BBS server-side software that I used in the 90s, I'm humbled by your offer and involvement here!).

Regarding **'+'**, I suppose we could use that NET0/1, and just "extend" the command set.  Given we're still in the review phase, it's just a one-character change.  If client or sever software lets you accumulate many commands (and the software inserts the `+`), then it's probably the better choice?

**Comment by kcgen** [comment]
@rswindell , 

Quick update: I'm now using the + command per the specification. It's the right thing to do as it follows the spec for any arbitrary multi-character command (which lots of modem vendors used and customized).  I've updated the PR description accordingly.

Longer story:

Since you mentioned + prefixing multi-character AT commands, I took another look at the code and noticed the "+" was being discarded prior to entering the AT command-parsing loop. 

This is because an even older version of the softmodem code wasn't properly handling the "+++" escape sequence to switch to command-mode, and ignoring it here was part of the prior implementation.  Fortunately we had a contributor fix the detection and handling of the "0.1s +++ 0.1s" sequence in commit: 4aee47a770dddcf74f4564fc25b48adcd7cfe943 

However, that fix didn't do away with the prior ignoring of the + in the command sequence. (so it turns out dosbox has never properly detected or handled multi-character command sequences).

A tiny change now allows + to be part of the AT command parsing loop, inside of which we now parse NET1 and NET0. I've also left a general comment in the "+" block to help future maintainers add more multi-character commands, if we like.

Here we see the parsing in action handling several toggles and echo on/off's:

**Comment by kcgen** [comment]
Having some fun.. communication is running clean even at 57600 Baud along with large file-transfers:

**Comment by dreamer** [comment]
Links to newer builds with all @kcgen's changes: [Linux](https://github.com/dosbox-staging/dosbox-staging/actions/runs/234527711), [Windows](https://github.com/dosbox-staging/dosbox-staging/actions/runs/234527713), [macOS](https://github.com/dosbox-staging/dosbox-staging/actions/runs/234527712)

**Comment by kcgen** [comment]
@rswindell - to help users, can you clarify when telnet-mode is needed?
Currently our documentation recommends always enabling telnet-mode for BBS-related activities (both client side and server-side).  But I found as a client, I can leave it disabled and still transfer files and see ANSI-BBS art without corruption. Thanks for any pointers! (I will updated the README accordingly).

**Comment by rswindell** [comment]
@kcgen, No problem. The short answer is: Telnet mode should be enabled whenever connecting to a telnet server.

Important Note: Not all BBSes on the Internet are behind telnet servers. Many of the TCP<->serial gateways in use by vintage BBSes (e.g. Commodore, Apple, Atari, etc.) are in fact not using telnet at all - they're just raw TCP gateways.

Telnet servers are normally listening on TCP port 23. A TCP server listening on any *other* TCP port may or may not be telnet. If the server is not listening on TCP port 23, that's a good clue that it may not be a telnet server.

You can usually connect to a raw TCP gateway/BBS using a telnet client, but will usually encounter issues such as:
- The "enter" key may do something unexpected on the BBS (e.g. ignored or treated as 2 ENTERs)
- File transfers using binary protocols (e.g. X/Y/ZMODEM) don't work most of the time

The reasons? In normal (non-binary) telnet mode, an enter key is expected to be sent by the client as either of the character-pairs: CR/LF (ASCII 13, 10) or CR/NUL (ASCII 13, 0). These special character pairs are translated by the server to the equivalent of a single "new-line" character/sequence for the software behind the server; for most BBSes, this should just a single CR (ASCII 13) character. A telnet client is never expected to send just a single CR character in normal (non-binary) mode. While in binary mode, this CR/LF and CR/NUL to CR translation between client and server does not occur. Telnet binary mode is normally automatically negotiated between the server and client before a file transfer and then disabled (back to normal/text mode) once the file transfer has ended. A non-telnet (raw) TCP server would see the CR/LF or CR/NUL sent by the client as two bytes and pass these to the underlying application software (e.g. a BBS) "as is" and that BBS might see and treat these as 2 different keystrokes. A non-telnet (raw) TCP server has no concept of binary and non-binary (text) telnet modes.

The reason that file transfers would fail on a non-Telnet BBS is that telnet clients and servers must always "escape" the character 255 (known as Telnet IAC). During file transfers, it is common to have the character 255 (0xFF) appear in the packets communicated between the sender and receiver and if any of those characters are not properly escaped/unescaped, then the receiver will detect a protocol error (e.g. checksum or CRC error) and usually retry the packet or abort the file transfer. At what point and how many IAC characters may appear in a file transfer is difficult to predict and would vary based on the contents of the file being transferred and the protocol used.

How can you tell if the BBS is or is not using Telnet? If you connect to a server using a raw TCP client (e.g. netcat) and can view the raw contents of the TCP stream (e.g. using Wireshark or 'netcat -o file'), then you can see if the server is attempting to negotiate using telnet commands, e.g.
```
< 00000067 ff fb 01 ff fb 03 ff fd 18 ff fd 20 ff fd 17 ff # ........... ....
< 00000077 fd 1f ff fd 27    

```
Those "ff xx xx" sequences are telnet commands sent by the server in an attempt to negotiate the settings of the telnet session  with the client (e.g. echo off, non-binary mode, etc.). If the server does not send any sequences like these, it is most likely not a telnet server.

Some telnet servers (e.g. Synchronet v3.18's terminal server) can automatically detect whether a TCP client is using telnet or not and behave accordingly for maximum compatibility and user-friendliness (and yes, that was a DOSbox inspired feature):
https://github.com/rswindell/sbbs/commit/40d829f2641c3e1aae76d89c16e84e9680a479cd

And of course, there are other non-telnet TCP protocols used by BBSes on the Internet as well: SSH and RLogin are prevalent, but WebSocket and telnets (telnet over TLS) are also possibilities.

**Comment by kcgen** [comment]
Thanks for the details explanation @rswindell.
This was pretty murky in the readme regarding the nullmodem and softmodem; I've tried to distill and clarrify in commit: https://github.com/dosbox-staging/dosbox-staging/pull/582/commits/120821f15551784e1d8f0423bd5a3674b56f9e88

**Comment by kcgen** [comment]
@dreamer, suggest we review for merge. I believe I've captured the user-actionable points per Rob's description in the README. 

The telnet enable/disable issue that this PR is meant to address is working as expected in all my tests. 

If/when Rob decides to try this, then we can address any additional issues when that time comes.

File: src/hardware/serialport/softmodem.h [hunk 1]
```diff
@@ -211,14 +211,14 @@ class CSerialModem : public CSerial {
 	std::unique_ptr<CFifo> tqueue;
 
 protected:
-	char cmdbuf[255] = {0};
+	char cmdbuf[353] = {0};
```
Comment (dreamer) [inline src/hardware/serialport/softmodem.h:214]: Is there any relation to magic number "253" added to `DoCommand` method?

File: README [hunk 2]
```diff
@@ -1327,23 +1330,69 @@ of the nullmodem connection. These are all parameters:
...
+ 1) Configure DOSBox's serial port as follows:
+
+    [serial]
+    serial1 = modem listenport:2323
 
-   [serial]
-   serial1 = modem listenport:2323
+    If you're positive that all the BBSes you plan to dial are hosted
+    behind Telnet servers, then enable telnet-mode as shown in step
+    1) in the BBS server section above (add telnet:1).
```
[the remark this answers is not in the document]
Comment (kcgen) [reply README:1366]: Yes, "in the BBS server section", which is the section above this client section :-)

But it will be clearer if I just repeat the content and drop the reference.

File: README [hunk 3]
```diff
@@ -1327,23 +1330,69 @@ of the nullmodem connection. These are all parameters:
...
+
+    [serial]
+    serial1 = modem telnet:1 listenport:2323
+
+    A port greater than 1024 allows you to run DOSBox as a normal user instead
+    of requiring root or administrator rights.
+
+ 2) Configure your machine to use a static IP address or be assigned a static
+    IP via DHCP, which can typically be configured in your router.
+
```
[the remark this answers is not in the document]
Comment (kcgen) [reply README:1351]: Thanks, yes, I will improve this!

File: include/support.h [hunk 4]
```diff
@@ -132,6 +132,15 @@ char * safe_strcat(char (& dst)[N], const char * src) noexcept {
...
 
+template<size_t N>
+int safe_strcmp(const char (& a)[N], const char * b) noexcept {
+	// Is 'b' at least as long as 'a'?
+	constexpr size_t N_without_null = N - 1;
+	if (strnlen(b, N) < N_without_null)
+		return -1;
+	return strncmp(a, b, N_without_null);
+}
+
```
[the remark this answers is not in the document]
Comment (kcgen) [reply include/support.h:143]: Yeah, I was just trying to programmatically get the needle-string length (instead of hardcoding), and also getting the string length of the haystack.   So even though `strnlen` is programmatically safe, it requires the developer to use it correctly - which is how it becomes unsafe in a "soft" sense. 

I will move it out of support and into the softmodem code; so other multi-char commands can use it.

File: src/hardware/serialport/softmodem.h (hunk 1 above)
Comment (kcgen) [reply src/hardware/serialport/softmodem.h:214]: The previous code didn't take the potential full length of a hostname into account (which is 253 characters). 

The AT command line can consist of a 99-character command sequence followed by D<phone/hostname>

So.. 99 + D + hostname can potentially hit 253 characters.

Comment (kcgen) [reply src/hardware/serialport/softmodem.h:214]: Added space for a terminating null character (now 354), and added a comment describing this.

File: include/support.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (kcgen) [reply include/support.h:143]: function moved.

File: README (hunk 3 above)
[the remark this answers is not in the document]
Comment (kcgen) [reply README:1351]: This wasn't clear enough; the intent is that the machine inside the network (behind the router/firewall) should have a static IP.  A lot of people runinng Windows just use DHCP and that makes a port-foward unreliable because typically the rule is IP-based (and not hostname-based); so it will work on day one, but 24-hours later their PC might get a new IP and then their BBS goes "offline". 

We don't want to go overboard (and out of scope) in the README, so I'll try to balance this out.

File: README (hunk 2 above)
[the remark this answers is not in the document]
Comment (kcgen) [reply README:1366]: Got it :-)

**Comment by kcgen** [comment]
@rswindell , thanks for reviewing the documentation.  Updates are inbound (dropped mentioning ANSI corruption, and clarified 8N1).

File: README [hunk 5]
```diff
@@ -1327,23 +1330,73 @@ of the nullmodem connection. These are all parameters:
...
+ 1) Configure DOSBox's serial port as follows:
+
+    [serial]
+    serial1 = modem telnet:1 listenport:2323
+
+    A port greater than 1024 allows you to run DOSBox as a normal user instead
+    of requiring root or administrator rights.
+
+ 2) Launch and configure your prefered DOS-based BBS software such as Renegade,
```
Comment (dreamer) [inline README:1349]: Typo: preferred

File: src/hardware/serialport/softmodem.cpp [hunk 6]
```diff
@@ -391,11 +395,22 @@ void CSerialModem::EnterConnectedState() {
 	CSerial::setRI(false);
 }
 
-void CSerialModem::DoCommand() {
+template <size_t N>
+bool starts_with(const char (&a)[N], const char *b) noexcept
```
Comment (dreamer) [inline src/hardware/serialport/softmodem.cpp:399]: Only a suggestion, does not block the review.

I suggest renaming this function to : `peek_token` or `peek_next_token` or `is_next_token` to communicate, that this function is used in context of parsing?

```
if (is_next_token("NET", scanbuf) {
```

It can stay as it - we just have `starts_with` function in support already, but it does not fulfil requirements for this use-case, I think.

**Review by dreamer — COMMENTED** [review]
@kcgen Last few nitpicky comments, otherwise it looks really nice :)

I would love to have confirmation from @rswindell that this implementation fulfils his requirements, but if he's busy we can merge it in and do some adjustments later on master (if necessary).

File: src/hardware/serialport/softmodem.cpp (hunk 6 above)
Comment (kcgen) [reply src/hardware/serialport/softmodem.cpp]: Thanks @dreamer; went with `is_next_token` because it more specifically describes what we're trying to do and that we're looking for token - as opposed to simply doing string compares.  Did one last test, and all is good.

**Review by dreamer — APPROVED** [review]
@kcgen Thanks :)

**Comment by milby79** [comment]
I've been following this with great interest, I was in the process of trying to restore my old BBS back from 1995.. it is Remote Access 2.02.

I had some initial success hosting it under linux using Dosbox, telnet clients can connect.. but the issue with file transfers failing on the CRC first block was present. After some searching I came across this... tonight I have compiled and built the staging release which contains the fix... after testing with both telnet:0 and telnet:1, I've come to the conclusion that Remote Access is indeed a TCP to serial host (uses a fossil driver), and as such won't benefit from the fix here?

So just to confirm, am I right in thinking that TCP to Serial hosts will not be able to use transfers such as Zmodem over telnet, in any form.

Thanks for all the efforts here by the way!

**Comment by kcgen** [comment]
> I've come to the conclusion that Remote Access is indeed a TCP to serial host (uses a fossil driver), and as such won't benefit from the fix here?

Thanks for digging into this @milby79; that's a good question and sounds like that's the case.

I've tested DOSBox's soft modem running terminal clients such as Terminate, Telemate, and Procomm connecting to Synchronet running natively server-side, and Zmodem runs clean for long transfers. 

Also tested DOSBox to DOSBox null modem connections with various DOS clients and servers (such as one Zmodem sending and the other receiving, as we all various games).

To understand the scenario you're exercising: your server-side is a real hardware DOS machine with Ethernet card and fossil driver running DOS BBS host software. You're then connecting to that using DOSBox soft modem and terminal software (and neither telnet:0 or telnet:1 is helping)?

**Comment by milby79** [comment]
Hi @kcgen, so the setup I have is as follows:

- Raspberry Pi 4, running Dietpi (debian), connected to my home Lan using RJ45.. on here I’m running Dosbox staging release.
- Dosbox is setup with 2 com ports, using the serial options of ‘modem telnet:1 listenport:2323’... or switching it to telnet:0.
- BBS software is Remote Access 2.02, using a fossil driver called FCI.exe.. the BBS is configured to run on com1. Modem initialising string is ATH0S0=1 E0... returns OK, and displays RING amd answers after the first ring when a telnet connection is attempted via port 2323.

- For the client, I’m using Netrunner under Win10-64, which in the connection details lets you select 
protocols of ssh, telnet... it’s set to telnet.. and this is connected to the same LAN as my BBS, the two devices can communicate with no issues etc.

When I run it with either telnet:1, or telnet:0, zmodem transfers fail and cycle restarting the first block with a CRC error...

I’ve also noticed since upgrading dosbox from 0.73-3 to staging, that when attempting to connect to the BBS over telnet, sometimes it’s saying connection failed and takes a few attempts before it makes a successful connection. I will test this further but it was something I noticed immediately, because I thought on my first attempt setting it to telnet:1 was somehow stopping it working completely, but through several attempts of switching this parameter I discovered it connected on both to the BBS, but sometimes not first time.

I had also observed when using telnet:1, in the terminal window which launches dosbox (which shows modem responses), it was giving some occasional errors about telnet IAC, 241 and 255 (from memory, not at my PC).

I also have a true hardware 486/100 running dos 6.22 and a serial to wifi modem adapter, I will test on this also (as the client dialling the BBS) using Terminate and see if this gives any different results... I’ve been using this to connect with other telnet BBS’s, and downloading files with zmodem with no issues, so I know this client is a good setup.

Hope all that info helps and adds some value to the situation, I hope to follow up with my 486 test, but I feel the issue is firmly on the BBS host side.. and that tcp to serial is the issue (if that describes my current setup).

Thanks,

Chris
